src/runtime/stopwatch.ts

```diff
@@ -67,16 +67,18 @@ export class PredictableTestRunStopwatch implements ITestRunStopwatch {
   }
 
   timestamp(): messages.ITimestamp {
+    const fakeTimestamp = this.convertToTimestamp(this.duration())
     this.count++
-    return convertToTimestamp(this.duration())
+    return fakeTimestamp
   }
-}
 
-function convertToTimestamp(duration: Duration): messages.ITimestamp {
-  const seconds = Math.floor(duration.seconds())
-  const nanos = Math.floor((duration.seconds() - seconds) * 1000000000)
-  return {
-    seconds,
-    nanos,
+  // TODO: Remove. It's impossible to convert timestamps to durations and vice-versa
```

Comment:
What is preventing us from doing this now?

Reply:
Nothing in particular - just didn't want to go down the rabbit hole. My thinking is that we need a more thorough review/overhaul of how we handle wall time vs monotonic time in a separate PR, and this can be addressed then.

Reply:
Okay yeah, so the stopwatch concept I guess becomes unnecessary and we should probably be able to infer any durations we want by finding the started/finished messages and using their timestamps.

(I'm happy to pick that up BTW since this is borne of my mistake.)

Reply:
We should use [monotonic](https://stackoverflow.com/questions/3523442/difference-between-clock-realtime-and-clock-monotonic) timestamps to calculate duration values and wall time to calculate timestamp values.

In JavaScript, wall time (`CLOCK_REALTIME`) is used in `Date.now()`. Monotonic time (`CLOCK_MONOTONIC`) is used in `performance.now()` (available on `window` in browsers and from `perf_hooks` in Node.js).

What's important to realise is that *wall time can go backwards*. In other words, `Date.now()` is unsuitable for durations. It's only suitable for timestamps. On the flip side, `performance.now()` is the right choice for durations, and the wrong choice for timestamps.

The only reason we're calculating the test run duration with wall time is that I found it difficult to do it with the current stopwatch. Now that I think of it, it's probably not that hard.

I think all of this will be easier if we drop the `duration` and `durations` modules and implement our own. It's also a good way to understand how to implement a stopwatch and a clock.

I propose the following API for a (wall) clock:

```TypeScript
type Timestamp = number
type Clock = () => Timestamp

const clock: Clock = ...
const timestamp = clock()
```

And a stop watch:

```TypeScript
type Duration = number
type StartStopwatch = () => StopStopwatch
type StopStopwatch = () => Duration

const startStopwatch: StartStopwatch = ...
const stopStopwatch = startStopWatch()
doSomeWork()
const duration = stopStopwatch()
```

I started this in #1430 but I abandoned it in favour of this PR because I went too far down a rabbit hole. I'd like to give it another try though. Or maybe you want to have a stab at it @davidjgoss?